File: src/main/java/com/everchanging/hierarchy/model/Employee.java
package com.everchanging.hierarchy.model;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.util.HashSet;
import java.util.Set;

@Data
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
@ToString(onlyExplicitlyIncluded = true)
public class Employee {
    @EqualsAndHashCode.Include
    @ToString.Include
    private final String name;

    private Employee supervisor;
    private Set<Employee> subordinates = new HashSet<>();

    public boolean addSubordinate(Employee employee) {
        return subordinates.add(employee);
    }
}

File: src/main/kotlin/com/everchanging/hierarchy/model/Employee.kt
package com.everchanging.hierarchy.model

data class Employee(val name: String) {
    var supervisor: Employee? = null
        private set
    var subordinates: MutableSet<Employee> = mutableSetOf()
        private set

    fun addSubordinate(subordinate: Employee): Boolean {
        subordinate.supervisor = this
        return subordinates.add(subordinate)
    }
}